packages/frameworks-qwik/src/index.ts

```diff
@@ -0,0 +1,382 @@
+ *   }
+ * };
+ *
+ * ```
+ *
+ * @module @auth/qwik
+ */
+
+import { Auth, skipCSRFCheck } from "@auth/core"
+import { AuthAction, Session } from "@auth/core/types"
+import { implicit$FirstArg, type QRL } from "@builder.io/qwik"
+import {
+  globalAction$,
+  routeLoader$,
+  z,
+  zod$,
+  type RequestEventCommon,
+} from "@builder.io/qwik-city"
+import { EnvGetter } from "@builder.io/qwik-city/middleware/request-handler"
+import { isServer } from "@builder.io/qwik/build"
+import { parseString, splitCookiesString } from "set-cookie-parser"
+import { GetSessionResult, QwikAuthConfig, qwikAuthActions } from "./types.js"
+
+export function qwikAuthQrl(
+  authOptions: QRL<(ev: RequestEventCommon) => QwikAuthConfig>
+) {
+  const useAuthSignIn = globalAction$(
+    async (
+      { providerId, callbackUrl: deprecated, options, authorizationParams },
+      req
+    ) => {
+      if (deprecated) {
+        console.warn(
+          "\x1b[33mWARNING: callbackUrl is deprecated - use options.callbackUrl instead\x1b[0m"
+        )
+      }
+      const { callbackUrl = deprecated ?? defaultCallbackURL(req), ...rest } =
+        options ?? {}
+
+      const isCredentials = providerId === "credentials"
+
+      const auth = await patchAuthOptions(authOptions, req)
+      const body = new URLSearchParams({ callbackUrl: callbackUrl as string })
+      Object.entries(rest).forEach(([key, value]) => {
+        body.set(key, String(value))
+      })
+
+      const baseSignInUrl = `/api/auth/${
+        isCredentials ? "callback" : "signin"
+      }${providerId ? `/${providerId}` : ""}`
+      const signInUrl = `${baseSignInUrl}?${new URLSearchParams(
+        authorizationParams
+      )}`
+
+      const data = await authAction(body, req, signInUrl, auth)
+
+      // set authjs.callback-url cookie. Fix for https://github.com/QwikDev/qwik/issues/5227
+      req.cookie.set("authjs.callback-url", callbackUrl, {
+        path: "/",
+      })
+
+      if (data.url) {
+        throw req.redirect(301, data.url)
+      }
+    },
+    zod$({
+      providerId: z.string().optional(),
+      callbackUrl: z.string().optional(),
+      options: z
+        .object({
+          callbackUrl: z.string(),
+        })
+        .passthrough()
+        .partial()
+        .optional(),
+      authorizationParams: z
+        .union([z.string(), z.custom<URLSearchParams>(), z.record(z.string())])
+        .optional(),
+    })
+  )
+
+  const useAuthSignOut = globalAction$(
+    async ({ callbackUrl }, req) => {
+      callbackUrl ??= defaultCallbackURL(req)
+      const auth = await patchAuthOptions(authOptions, req)
+      const body = new URLSearchParams({ callbackUrl })
+      await authAction(body, req, `/api/auth/signout`, auth)
+    },
+    zod$({
+      callbackUrl: z.string().optional(),
+    })
+  )
+
+  const useAuthSession = routeLoader$((req) => {
+    return req.sharedMap.get("session") as Session | null
```

Comment:
I think this has to be implemented similarly to signin/signout actions, with the call to `Auth` core - unless Im missing something here. the implementation of this function is currently in `getSessionData`.

Reply:
`onRequest` is reading the session from the Auth core and it's setting it the `sharedMap`.
onRequest is running before the useSession so we don't need to call the Auth core another time

Reply:
ah I see, the way the `onRequest` middleware works is that if we export it in `plugin@auth.ts`, it would be ran before anything.

I think this could not be a good default for `onRequest` - instead of calling the `getSessionData` function on every requests that are not auth,  how about we put the function in `useSession` and allow apps to lazily call it when needed? It would be similar to the SvelteKit approach, where we expose a `getSession` instead of the session object. cc @balazsorban44